Refactor mage output

austinbyers opened the pull request:
## Background

Incorporating feedback from developers and users to show more useful and human-readable output, improve error handling, and more.

I'd encourage checking out the branch and running `mage fmt test:ci` to get a sense of the new style! Then try in debug mode: `mage test:go` vs `mage -v test:go`

## Changes

### General
* All `mage` commands print to the console using a `zap` logger instead of `fmt.Print`. This has a number of benefits:
    * Proper logging levels - we can say `logger.Debugf()` instead of `if mg.Verbose() { fmt.Printf() }`
    * Consistency - we already use the `zap` logger in all Golang Lambda functions
    * Clarity - I've configured the logger to show colored status levels, time to the minute, and file/line number (in debug mode only)
    * It's now much easier to distinguish output printed directly by Panther from shell commands
* All errors have been annotated with context. For example, `return fmt.Errorf("failed to describe stack %s: %v", stack, err)"` It was not always clear where error messages came from until now
    * I also now sometimes simplify code in higher-level functions by failing when we encounter an error instead of returning it all the way up the call stack. Since this is like a CLI, most errors are fatal and we can just log them and stop.
* Use `filepath` consistently instead of `path` - they are two different packages with almost identical interfaces, but `filepath` should always be used for interacting with files

### Individual Commands
* `build:lambda` stops immediately if the source hasn't changed since the last compilation (saves ~20 seconds)
* `deploy` 
    * Fail immediately if something is wrong with the user environment (unsupported AWS region, Go out of date, docker unavailable)
    * The graphQL is uploaded by `aws cloudformation package` instead of us injecting it manually. We may get rid of the AWS CLI at some point, but until then there's no reason not to use it for all the resources it supports.
   * Eliminate a few redundant `DescribeStack` calls - we already have the stack outputs when waiting on a change set, we don't need to query the stack again after it finishes 
* `fmt` will also format CloudFormation and the web source
    * bugfix - generated API files are always formatted the same way
* `setup` is broken into a few different commands so the CI jobs can install only what they need
    * Developers can now run `mage setup:dev` to install everything, including web
* `test:lint` and `test:unit` have been replaced with test suites for each language to more accurately describe the developer workflow
    * `test:go`, `test:python`, `test:web`, `test:cfn` let you test changes for only the type of files you are working with. This will also let us break CI into these 4 separate checks, each of which requires different setup
    * go unit test output is filtered by default to only show failures (it's getting very noisy)
    * This adds web tests to `mage` for the first time
* `test:integration` is many times faster now - I only test packages with an `integration_test.go` file instead of using `go test` to check the entire project


## Testing

- Add unit tests for swagger API embedding
- `mage test:ci`

```
22:09:28	INFO	test:cfn: cfn-lint 28 templates
22:09:33	INFO	test:go: unit tests
22:09:39	INFO	test:go: golangci-lint
22:09:41	INFO	test:python: unit tests
----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
----------------------------------------------------------------------
Ran 32 tests in 0.028s

OK
----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
22:09:42	INFO	test:python: bandit security linting
56 [0.. 50.. ]
22:09:42	INFO	test:python: pylint
22:09:47	INFO	test:python: mypy type-checking
22:09:48	INFO	test:web: npm run lint

> panther@1.0.0 lint /Users/austin/go/src/github.com/panther-labs/panther
> eslint "{web/src/**,.}/*.{ts,js,tsx}" && prettier --list-different "{.circleci,api,build,deployments}/**/*.yml" && prettier --list-different "{web/src/**,.}/*.{ts,js,tsx,md,json,yml}"

22:10:05	INFO	test:web: npm audit

                       === npm audit security report ===

found 0 vulnerabilities
 in 35862 scanned packages
22:10:07	INFO	test:ci: PASS
```

- `mage deploy`

```
21:33:10	INFO	build:api: generating Go SDK for 4 APIs (api/gateway/*/api.yml)
21:33:10	INFO	build:lambda: compiling 20 Go Lambda functions (internal/.../main)
21:33:17	INFO	deploy: UPDATE CloudFormation stack panther-buckets (may take several minutes)
21:33:18	INFO	deploy: packaging deployments/backend.yml assets in s3://panther-source-101802775469-us-west-2 (may take several minutes)

Successfully packaged artifacts and wrote output template to file out/deployments/package.backend.yml.
Execute the following command to deploy the packaged template
aws cloudformation deploy --template-file /Users/austin/go/src/github.com/panther-labs/panther/out/deployments/package.backend.yml --stack-name <YOUR STACK NAME>
21:33:41	INFO	deploy: UPDATE CloudFormation stack panther-app (may take several minutes)
21:36:11	INFO	deploy: logging in to remote image repo
WARNING! Your password will be stored unencrypted in /Users/austin/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

21:36:12	INFO	deploy: docker build web server (deployments/web/Dockerfile)
21:36:13	INFO	deploy: pushing docker image to remote repo
The push refers to repository [101802775469.dkr.ecr.us-west-2.amazonaws.com/panther-web]
009834d1c003: Layer already exists
020caab22792: Layer already exists
e5c24028b9b6: Layer already exists
2d3216389d53: Layer already exists
a1f06f230492: Layer already exists
20d6ac69de87: Layer already exists
15b7d94033f5: Layer already exists
4322be3f308a: Layer already exists
77cae8ab23bf: Layer already exists
6eca55bcc93913e9709304d467f9969c1d9ae277dffe1129184027fa9ffd5801: digest: sha256:a38b65609abdc6fc45f8547339aaa45e51c06cc73c2ffbade6ef2e627ed34136 size: 2203
21:36:15	INFO	deploy: packaging deployments/frontend.yml assets in s3://panther-source-101802775469-us-west-2 (may take several minutes)

Successfully packaged artifacts and wrote output template to file out/deployments/package.frontend.yml.
Execute the following command to deploy the packaged template
aws cloudformation deploy --template-file /Users/austin/go/src/github.com/panther-labs/panther/out/deployments/package.frontend.yml --stack-name <YOUR STACK NAME>
21:36:16	INFO	deploy: UPDATE CloudFormation stack panther-app-frontend (may take several minutes)
21:36:31	INFO	deploy: finished successfully in 3m22.281320321s
```

- `mage test:integration`

```
22:06:20	INFO	build:api: generating Go SDK for 4 APIs (api/gateway/*/api.yml)
22:06:20	INFO	test:integration: go test ./internal/compliance/compliance_api/main -run=TestIntegration*
22:06:27	INFO	test:integration: go test ./internal/compliance/resources_api/main -run=TestIntegration*
22:06:31	INFO	test:integration: go test ./internal/compliance/snapshot_api/main -run=TestIntegration*
22:06:40	INFO	test:integration: go test ./internal/core/analysis_api/main -run=TestIntegration*
22:06:48	INFO	test:integration: go test ./internal/core/organization_api/main -run=TestIntegration*
22:06:52	INFO	test:integration: go test ./internal/core/outputs_api/main -run=TestIntegration*
22:06:56	INFO	test:integration: python policy engine
.
----------------------------------------------------------------------
Ran 1 test in 0.257s

OK
```

nhakmiller commented:
Not sure where in the code this issue manifests, but if you haven't run `npm i` yet running `mage test:ci` just vomits tens of thousands of lines of output at you. Perhaps we're just ok with this if we're expecting users to follow our setup instructions before running anything, but just wanted to throw it out there.

austinbyers commented:
> Not sure where in the code this issue manifests, but if you haven't run npm i yet running mage test:ci just vomits tens of thousands of lines of output at you.

Good feedback! I added `npm i` to the `mage setup` command and also added a check for that condition since the error output is indeed confusing

kostaspap commented on tools/mage/logger.go:
```diff
@@ -0,0 +1,69 @@
+	// If the error is returned from a mage command, mage will print it, but without our logging format.
+	// So we log the error ourselves and exit manually for cleaner output.
+	os.Exit(1)
+}
```
You don't need this method, SugaredLogger `Fatal` logging already has this behavior. From docs
```
// Fatal uses fmt.Sprint to construct and log a message, then calls os.Exit.
func (s *SugaredLogger) Fatal(args ...interface{}) {
	s.log(FatalLevel, "", args, nil)
}
```

austinbyers replied:
I looked for something like that, not sure how I missed it the first time

austinbyers replied:
This is also better because it will keep the proper line number for the error messages

kostaspap commented on tools/mage/test_namespace.go:
```diff
@@ -49,129 +48,209 @@ var (
+	logger.Info("test:python: bandit security linting")
+	args = []string{"--recursive"}
 	if mg.Verbose() {
 		args = append(args, "--verbose")
```
See comment I made prior. You could be checking the logger levels instead of the `mg.Verbose` flag since we use log levels everywhere in the code.

kostaspap commented on tools/mage/test_namespace.go:
```diff
@@ -49,129 +48,209 @@ var (
-	args := []string{"test", "-cover", "./..."}
-	if mg.Verbose() {
-		args = append(args, "-v")
+// Web Lint web source
```
This is meant to run tests as well, not just linting right?

austinbyers replied:
True, there aren't tests yet but there will be

nhakmiller commented on tools/mage/deploy_certificate.go:
```diff
@@ -189,41 +141,48 @@ func generateKeys() error {
-	err = os.MkdirAll(keysDirectory, certFilePermissions)
-	if err != nil {
-		return err
+	if err = os.MkdirAll(keysDirectory, certFilePermissions); err != nil {
```
Part of the reason we did it like this is so that the keys are not re-generated every time someone rebuilds project, although I suppose the small amount of time it takes to generate a 2048 bit key might not warrant that, especially since we're already re-using the ACM/IAM certificate if it already exists. I would be ok either way on this.

austinbyers replied:
No, I think `keys` should be separate, the way they are now:

- They don't have to be autogenerated, users can in theory provide their own certificate
- `mage clean` would destroy your private keys, which is not what people would expect

kostaspap replied:
Sorry, maybe I am missing something out... Does the current flow allow the users to provide their own certificate? I mean, we create the "keys" folder and put the keys in there. How can the users make use of that folder to upload the certificate? 
And why would users want to have the keys auto-generated keys in the laptop/docker container that they used to deploy panther? Cert is uploaded to ACM anyway.

nhakmiller replied:
So the `keys` directory is a bit weird and we support a few workflows for configuring the certificate.

1. Users provide us nothing. In this case, we generate the keys, store them, upload the certificate, and then in the future use the uploaded certificate.
2. Users provide an ARN to a certificate in `deployments/panther_config.yml`. In this case, we generate nothing and just pass the ARN through to the stacks that need it.
3. Users place their own pre-generated private keys in the `keys` directory. In this case, we upload the provided keys to AWS and create a certificate, then use that certificate in the future. This is useful for organizations that do not have a big AWS presence (perhaps using us for log analysis primarily) so they don't want to go to ACM themselves but do have an internal CA that cuts certs for them that their hosts trust.

The primary reason we persist the `keys` directory is for case 3.

The secondary reason we persist the `keys` directory is a micro optimization; when people are doing constant fresh deployments (like us) they don't have to wait for the random key to be generated each time. Very minor reason.

We can always just drop support for case 3, and say anyone that wants to provide their own certs/keys goes through case 2. Or we could say we support case 3, but if you run mage clean you have to go find the keys/cert file again.

austinbyers commented on docs/gitbook/development.md:
```diff
@@ -26,7 +26,7 @@ Install [Docker](https://docs.docker.com/install/) and make sure the daemon is r
 Install [Mage](https://magefile.org/#installation): `go get github.com/magefile/mage`
 
-Finally, install the remaining development libraries with `mage setup && npm i`
+Finally, install the remaining development libraries with `mage setup:dev`
```
Sure, I like `setup:all`

austinbyers commented on tools/mage/fmt.go:
```diff
@@ -36,61 +34,66 @@ var (
-	if err := sh.Run("gofmt", append([]string{"-l", "-s", "-w"}, goTargets...)...); err != nil {
-		return err
+	// cloudformation
+	logger.Info("fmt: prettier deployments")
```
yeah, it's the directory, I can add the trailing slash

austinbyers replied:
I also updated the error message

austinbyers commented on tools/mage/fmt.go:
```diff
@@ -36,61 +34,66 @@ var (
+	// web
+	logger.Info("fmt: prettier web")
+	args = []string{"--write", "{web/src/**,.}/*.{ts,js,tsx,md,json,yml}"}
+	if !mg.Verbose() {
```
> no mg.Verbose flag

We use `mg.Verbose` to determine the log level at the very beginning, so at the end of the day it's the same thing

I think we should leave it as is - the logging library is independent of how we invoke commands

austinbyers commented:
I verified I was able to deploy again from scratch
